### ISSUE TITLE
Segfault building intro-vegan vignette on multiple OSes

### ISSUE THREAD

**Issue by gavinsimpson** [vegandevs/vegan#572]
`R CMD  build` is failing on the GitHub actions workflows with a segfault or similar when building the intro-vegan vignette. For example, this is on Ubuntu (latest LTS, R-release):

https://github.com/vegandevs/vegan/actions/runs/4676977547/jobs/8283939237#step:12:63

and same for MacOS X (latest OS version, R-release):

https://github.com/vegandevs/vegan/actions/runs/4676977547/jobs/8283938828#step:12:63

Windows seems to be failing at the same place but the output is missing.

What the CI is seeing on multiple OSes is:

```
--- re-building ‘intro-vegan.Rnw’ using Sweave
 *** caught segfault ***
address 0x5[62](https://github.com/vegandevs/vegan/actions/runs/4676977547/jobs/8283939237#step:12:64)c00000001, cause 'memory not mapped'
Traceback:
 1: getF(permutations, E = E, Q = Q, QZ = QZ, effects = effects,     w = w, first = first, isPartial = isPartial, isCCA = isCCA,     isDB = isDB, q = q, r = r)
 2: permutest.cca(object, permutations = permutations, model = model,     parallel = parallel, ...)
 3: anova.cca(ord)
 4: anova(ord)
 5: eval(expr, .GlobalEnv)
 6: eval(expr, .GlobalEnv)
 7: withVisible(eval(expr, .GlobalEnv))
 8: doTryCatch(return(expr), name, parentenv, handler)
 9: tryCatchOne(expr, names, parentenv, handlers[[1L]])
10: tryCatchList(expr, classes, parentenv, handlers)
11: tryCatch(expr, error = function(e) {    call <- conditionCall(e)    if (!is.null(call)) {        if (identical(call[[1L]], quote(doTryCatch)))             call <- sys.call(-4L)        dcall <- deparse(call, nlines = 1L)        prefix <- paste("Error in", dcall, ": ")        LONG <- [75](https://github.com/vegandevs/vegan/actions/runs/4676977547/jobs/8283939237#step:12:77)L        sm <- strsplit(conditionMessage(e), "\n")[[1L]]        w <- 14L + nchar(dcall, type = "w") + nchar(sm[1L], type = "w")        if (is.na(w))             w <- 14L + nchar(dcall, type = "b") + nchar(sm[1L],                 type = "b")        if (w > LONG)             prefix <- paste0(prefix, "\n  ")    }    else prefix <- "Error : "    msg <- paste0(prefix, conditionMessage(e), "\n")    .Internal(seterrmessage(msg[1L]))    if (!silent && isTRUE(getOption("show.error.messages"))) {        cat(msg, file = outFile)        .Internal(printDeferredWarnings())    }    invisible(structure(msg, class = "try-error", condition = e))})
12: try(withVisible(eval(expr, .GlobalEnv)), silent = TRUE)
13: evalFunc(ce, options)
14: tryCatchList(expr, classes, parentenv, handlers)
15: tryCatch(evalFunc(ce, options), finally = {    cat("\n")    sink()})
16: driver$runcode(drobj, chunk, chunkopts)
17: utils::Sweave(...)
18: engine$weave(file, quiet = quiet, encoding = enc)
19: doTryCatch(return(expr), name, parentenv, handler)
20: tryCatchOne(expr, names, parentenv, handlers[[1L]])
21: tryCatchList(expr, classes, parentenv, handlers)
22: tryCatch({    engine$weave(file, quiet = quiet, encoding = enc)    setwd(startdir)    output <- find_vignette_product(name, by = "weave", engine = engine)    if (!have.makefile && vignette_is_tex(output)) {        texi2pdf(file = output, clean = FALSE, quiet = quiet)        output <- find_vignette_product(name, by = "texi2pdf",             engine = engine)    }    outputs <- c(outputs, output)}, error = function(e) {    thisOK <<- FALSE    fails <<- c(fails, file)    message(gettextf("Error: processing vignette '%s' failed with diagnostics:\n%s",         file, conditionMessage(e)))})
23: tools::buildVignettes(dir = ".", tangle = TRUE)
An irrecoverable exception occurred. R is aborting now ...
Segmentation fault (core dumped)
Error: Error in proc$get_built_file() : Build process failed
```

As there have been changes in the permutation test C code in response to Cajo's recent paper (I believe you checked those changes in @jarioksa ?), this could indicate an issue there that wasn't identified in local tests.

I will see if I can confirm the problem using the CRAN macos build machine and CRAN's win-builder, as well as running checks locally myself on Ubuntu.

**Comment by jarioksa** [comment]
Segfaults fixed in commit 28cf00f.

**Comment by gavinsimpson** [comment]
Checks are issue-free; closing

**Comment by jarioksa** [comment]
I was a grumpy old man – again. When you look at the fix, it was obvious, but it was not easy to find. All my own continuous tests had been successful, but I connected to our Linux server which showed the error and bisected the code. Bisection gave as a first bad commit commit a commit that itself was OK (742cfa9a) but it was called in another function a few lines before the faulty piece. In hindsight this fix was obvious: now I remember that I had thought earlier that this very piece of code needs changing. However I had a long break in developing, and forgot about this, and when the tests were successful I just merged the branch.
